Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,167 @@
+    sampler: An instance of one of the :meth:`~optimisation.utilities.sampler` classes which has a `next` function implemented and a `num_indices` property.
+        This sampler is called each time gradient is called and  sets the internal `function_num` passed to the `approximate_gradient` function.  The `num_indices` must match the number of functions provided. Default is `Sampler.random_with_replacement(len(functions))`. 
+    
+    Note
+    ------
+    
+    The user has the option of calling the class method `warm_start_approximate_gradients` after initialising this class. This will compute and store the gradient for each function at an initial point. If this method is not called, the gradients are initialised with zeros. 
+
+    References
+    ----------
+    Schmidt, M., Le Roux, N. and Bach, F., 2017. Minimizing finite sums with the stochastic average gradient. Mathematical Programming, 162, pp.83-112. https://doi.org/10.1007/s10107-016-1030-6
+
+    """
+
+    def __init__(self, functions,  sampler=None):
+        self._list_stored_gradients = None
+        self._full_gradient_at_iterate = None
+        self._warm_start_just_done=False
+        self._stoch_grad_at_iterate=None
+        
+        super(SAGFunction, self).__init__(functions, sampler)
+
+
+        
+
+    def approximate_gradient(self, x, function_num,  out=None):
+        """ Returns the gradient of the selected function or batch of functions at :code:`x`. 
+            The function_num is selected using the sampler. 
+
+        Parameters
+        ----------
+        x: element in the domain of the `functions`
+
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+
+        """
+        if self._list_stored_gradients is None: 
+            self._list_stored_gradients = [
+                0*x for fi in self.functions]
+            self._full_gradient_at_iterate = 0*x
+            
+
+        self._stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        
+        self._update_data_passes_indices([self.function_num])
+
+        self._stochastic_grad_difference = self._stoch_grad_at_iterate.sapyb(
+            1., self._list_stored_gradients[function_num], -1.)
+
+   
+        out =self._update_approx_gradient(out)
+
+        self._list_stored_gradients[function_num].fill(
+            self._stoch_grad_at_iterate)
+        self._full_gradient_at_iterate.sapyb(
+            1., self._stochastic_grad_difference, 1., out=self._full_gradient_at_iterate)
+
+        return out
+    
+    def _update_approx_gradient(self, out):
+        """Internal function used to differentiate between the SAG and SAGA calculations"""
+        if out is None:
+            # due to the convention that we follow: without the 1/n factor
```

Comment:
Has this convention been stated anywhere? Make sure it has and consider referring to that place from here.